src/main/java/com/redhat/rhjmc/containerjfr/net/TargetConnectionManager.java

```diff
@@ -93,56 +97,55 @@
      * finished with it. When possible, clients should use executeConnectedTask instead, which does
      * perform automatic cleanup when the provided task has been completed.
      */
-    public JFRConnection connect(String targetId) throws Exception {
+    public JFRConnection connect(ConnectionDescriptor connectionDescriptor) throws Exception {
         try {
-            return attemptConnectAsJMXServiceURL(targetId);
+            return attemptConnectAsJMXServiceURL(connectionDescriptor);
+        } catch (MalformedURLException mue) {
+            return attemptConnectAsHostPortPair(connectionDescriptor);
         } catch (Exception e) {
-            return attemptConnectAsHostPortPair(targetId);
+            throw e;
         }
     }
 
-    private JFRConnection attemptConnectAsJMXServiceURL(String url) throws Exception {
-        return connect(new JMXServiceURL(url));
+    private JFRConnection attemptConnectAsJMXServiceURL(ConnectionDescriptor connectionDescriptor)
+            throws Exception {
+        return connect(
+                new JMXServiceURL(connectionDescriptor.getTargetId()),
+                connectionDescriptor.getCredentials());
     }
 
-    private JFRConnection attemptConnectAsHostPortPair(String s) throws Exception {
+    private JFRConnection attemptConnectAsHostPortPair(ConnectionDescriptor connectionDescriptor)
+            throws Exception {
+        String s = connectionDescriptor.getTargetId();
         Matcher m = HOST_PORT_PAIR_PATTERN.matcher(s);
         if (!m.find()) {
-            return null;
+            throw new MalformedURLException(s);
         }
         String host = m.group(1);
         String port = m.group(2);
         if (port == null) {
             port = "9091";
         }
-        return connect(host, Integer.parseInt(port));
+        return connect(
+                new JMXServiceURL(
+                        "rmi", "", 0, String.format("/jndi/rmi://%s:%s/jmxrmi", host, port)),
```

Comment:
Real minor thing, but this would be more readable IMO if we used JMC's ConnectionToolkit.createServiceURL. Maybe worth adding a wrapper for it in core next time you make changes there.

Reply:
That's a good call. I'll file a -core issue so it doesn't get forgotten.

Reply:
https://github.com/rh-jmc-team/container-jfr-core/issues/44